### ISSUE TITLE
Strange eval behaviour

### ISSUE THREAD

**Issue by joncys** [omphalos/thermite#2]
Hi, I'm dissecting the code to get the grasp on how it works.
I'm running the example code (client side via `browserify`:

```
var thermite = require('thermite');

var hotSwappableCode = thermite.eval('(' + function() {
  document.addEventListener("mousemove", function(evt) {
    console.log(evt.clientX * evt.clientY)
  })
} + ')()')

hotSwappableCode.hotSwap('(' + function() {
  document.addEventListener("mousemove", function(evt) {
    console.log(evt.clientX + evt.clientY)
  })
} + ')()')
```

I've noticed that whenever I move my mouse, the code get's executed twice – one time reevaluated and the other time from the cache, I've added some logging into the template wrapper:

```
var template = ['function $thermiteTemplate() {',
  '',
  'if(__thermiteMap.$thermiteContextID.$thermiteBlockID.codeVersion >',
  '  (__thermiteFunctionVersion_$thermiteBlockID || -1)) {',
  '',
  '  __thermiteFunction_$thermiteBlockID =',
  '    eval(__thermiteMap.$thermiteContextID.$thermiteBlockID.code)',
  '  console.log("EVAL HAS RUN");',
  'console.log(__thermiteMap.$thermiteContextID.$thermiteBlockID.code)',
  '',
  '  __thermiteFunctionVersion_$thermiteBlockID =',
  '    __thermiteMap.$thermiteContextID.$thermiteBlockID.codeVersion',
  '}',
  'console.log("EVAL NOT RUN");',
  'console.log(__thermiteMap.$thermiteContextID.$thermiteBlockID.code);',
  '',
  'return __thermiteFunction_$thermiteBlockID.apply(this, arguments)',
  '',
'}'].join('\n')
```

And here's the output I'm getting:

<img width="444" alt="screen shot 2015-09-28 at 22 22 27" src="https://cloud.githubusercontent.com/assets/373939/10146079/89478a64-662f-11e5-8f68-f416c8f25881.png">

Maybe you have some insights into what's happening here, from the general overview, I would think the event handler should always be running from the cache, but it seems that the `functionVersion` variable for that particular function remains `undefined` and to me it seems, that it should be wrapped into a IIFE, but the wrapping IIFE never gets executed. But I may be getting something totally wrong here.

Any help is very much appreciated, thank you!

**Comment by omphalos** [comment]
@joncys Thank you very much for this bug report!  Your analysis was very thorough and your debugging steps helped me track it down very quickly.

Okay, so first of all I'd like to clarify that the function is not executed twice.  If you look inside the template, `EVAL HAS RUN` will happen when the code is `eval`'ed -- the bug is that this shouldn't happen every time in a function expression (it works properly for function declarations).  `EVAL NOT RUN` is a little misleading -- if you look closely `apply` is only called in one place.  The code is supposed do something like:
1. Check for a newer version.  If it's newer, `eval` it.  This step creates our function reference, and makes sure it's the latest version of the function.  `eval` doesn't actually call the function - it just defines it.
2. Whether there was a new version or not, we want to `apply` it.  This step actually invokes the function.

> it should be wrapped into a IIFE

Absolutely correct - that's the fix.  Actually it looks like there is a similar issue with recursive function expressions here that I need to fix as well.  In case you're curious, it's the `addVersionToExpression` code that needs to be fixed with an IIFE.

Again, I appreciate the excellent analysis and bug report.  This is fixed now in 0.0.5.

**Comment by joncys** [comment]
> Okay, so first of all I'd like to clarify that the function is not executed twice.

@omphalos Yeah, I absolutely worded that incorrectly all the way. The function gets executed only once, but is evaled every time. I was too long into reading the source code. :)

:+1: Thanks for the fix, very interested to see the diff! And thanks for writting this thingy!
